File: max/engine/phrase_engine.js
"use strict";

/**
 * phrase_engine.js — sequence-level phrase generation (the SEQUENCE is the unit).
 *
 * Faithful port of python/src/engines/phrase_engine.py. Answers "give me an
 * N-bar phrase in key K" as a list of [chordSymbol, durationBeats] pairs whose
 * harmonic rhythm is LEARNED (a semi-Markov chain over ROOT MOTION with an
 * explicit, metric-position-conditioned duration model) and whose ending is an
 * imposed V(7) -> I cadence.
 *
 * WHY a port rather than a rewrite: the Python engine is the reference. Its
 * draw ORDER and distribution access (including every backoff fallthrough) are
 * reproduced verbatim so the JS output matches the corpus statistics. The one
 * intentional divergence is the PRNG: Max has no seedable RNG, so rng.js's
 * mulberry32 replaces Python's random.Random. Identical seeds therefore do NOT
 * produce identical sequences across the two languages — only the same
 * DISTRIBUTIONS do, which is why the cross-check compares histograms, never
 * individual draws.
 */

const fs = require("fs");
const { CANON_ROOT, QUALITY_INTERVALS, PITCH_CLASSES, parseKey } = require("./chord_vocab.js");
const { makeRng } = require("./rng.js");

// Rejection-sample this many exact-length runs before falling back to a
// truncated draw that always fits (see run()).
const MAX_TRIES = 200;
// Beats reserved at the end for the tonic resolution (one 4/4 bar).
const CADENCE_BARS = 1;
// Multiplicative pull toward in-key roots at full Cadence. At gravity=0 the
// corpus transition distribution is reproduced untouched.
const MAX_TONAL_PULL = 4.0;

// Scale degrees (semitones from the tonic) treated as "at home" in each mode.
// Minor includes the raised 7th (11) so the dominant belongs. Stored as Sets
// for O(1) membership; degrees are always non-negative (via mod12).
const DIATONIC = {
  maj: new Set([0, 2, 4, 5, 7, 9, 11]),
  min: new Set([0, 2, 3, 5, 7, 8, 10, 11]),
};

// JS `%` can be negative for negative operands; Python's cannot. Every modulo
// on a possibly-negative pitch value must go through this to match the data's
// key-space arithmetic. (mod12 is not exported by chord_vocab, so reproduced.)
function mod12(x) {
  return ((x % 12) + 12) % 12;
}

/**
 * isDiatonic(rootPc, quality, tonicPc, mode, home) -> boolean.
 *
 * Are ALL of this chord's notes in the key's scale? Testing the root alone is
 * WRONG: in A minor, A:maj7 has a diatonic ROOT but a raised third and seventh.
 * Testing every interval of the quality yields exactly the diatonic sevenths
 * (in C: Cmaj7 Dm7 Em7 Fmaj7 G7 Am7 Bhdim7) and correctly rejects A:maj7 in A
 * minor. `home` is the DIATONIC Set for `mode`.
 */
function isDiatonic(rootPc, quality, tonicPc, mode, home) {
  const degree = mod12(rootPc - tonicPc);
  const intervals = QUALITY_INTERVALS[quality];
  if (!intervals || intervals.length === 0) return false;
  for (const iv of intervals) {
    if (!home.has(mod12(degree + iv))) return false;
  }
  return true;
}

/**
 * pick(rng, dist) — sample a key from a [[key, prob], ...] table.
 * r = rng(); walk cumulative prob; return the first key whose bucket covers r,
 * else the last key (guards against probabilities summing just under 1).
 */
function pick(rng, dist) {
  const r = rng();
  let acc = 0;
  for (let i = 0; i < dist.length; i++) {
    acc += dist[i][1];
    if (r <= acc) return dist[i][0];
  }
  return dist.length ? dist[dist.length - 1][0] : null;
}

function clamp01(x) {
  const n = Number(x);
  if (!Number.isFinite(n)) return 0;
  return Math.max(0, Math.min(1, n));
}

/**
 * createPhraseEngine(modelJsonPath, rng) -> { generate(...) }.
 *
 * `rng` is the default stream (a function -> float in [0,1)); when generate()
 * is given a numeric `seed` it makes a fresh mulberry32 stream instead. Throws
 * if the model file is missing or malformed.
 */
function createPhraseEngine(modelJsonPath, rng) {
  let m;
  try {
    m = JSON.parse(fs.readFileSync(modelJsonPath, "utf8"));
  } catch (exc) {
    throw new Error(`cannot load phrase model ${modelJsonPath}: ${exc.message}`);
  }
  for (const required of ["transition", "duration", "cadence", "duration_support"]) {
    if (!(required in m)) {
      throw new Error(`phrase model ${modelJsonPath} missing '${required}'`);
    }
  }

  const trans = m.transition; // {quality: [[[dpc, nextQ], p], ...]}
  const transMarginal = m.transition_marginal; // [[[dpc, q], p], ...]
  const dur = m.duration; // {"quality|m": [[dur, p], ...]}  (keys contain '|', safe as object)
  const durAll = m.duration_all; // [[dur, p], ...]
  const qualities = m.qualities || [];
  const cadence = m.cadence;
  const support = m.duration_support;

  // duration_backoff is keyed by integer metric position ("0".."3"). Integer
  // keys in a plain JS object get reordered on iteration; we only ever do a
  // direct lookup here so order is immaterial, but the project rule is that
  // integer-keyed collections live in a Map — so honor it.
  const durBackoff = new Map();
  for (const k of Object.keys(m.duration_backoff || {})) {
    durBackoff.set(Number(k), m.duration_backoff[k]);
  }

  // onset_position has INTEGER keys and MUST be a Map (a plain object would
  // reorder the metric positions). [[m, p], ...] -> Map<number, number>.
  const onsetPos = new Map();
  for (const [k, p] of m.onset_position || []) {
    onsetPos.set(Number(k), p);
  }

  // Memoize isDiatonic on (degree, quality, mode): the transition reweighting
  // calls it for every candidate on every step, and the truth value depends
  // only on the degree (not the absolute root) and the mode.
  const diatonicMemo = new Map();
  function isDiatonicMemo(rootPc, quality, tonicPc, mode, home) {
    const degree = mod12(rootPc - tonicPc);
    const memoKey = degree + "|" + quality + "|" + mode;
    let hit = diatonicMemo.get(memoKey);
    if (hit === undefined) {
      hit = isDiatonic(rootPc, quality, tonicPc, mode, home);
      diatonicMemo.set(memoKey, hit);
    }
    return hit;
  }

  /**
   * drawMotion(quality, rng, root, ctx) -> [dpcInt, nextQuality].
   *
   * TONAL GRAVITY: trained key-free, an order-1 motion chain reproduces the
   * corpus's constant modulation, so a long phrase drifts from home. With the
   * key known at generation, ctx.gravity>0 multiplies each candidate's weight
   * by `bonus` when the resulting chord is wholly diatonic, then renormalizes.
   * Multiplicative, so out-of-key colour stays reachable — just rarer.
   */
  function drawMotion(quality, rngFn, root, ctx) {
    let dist = trans[quality] || transMarginal;
    if (!dist || dist.length === 0) return [5, "7"]; // descending fifth into a dominant
    if (ctx && ctx.gravity > 0) {
      const bonus = 1.0 + MAX_TONAL_PULL * Math.min(1.0, ctx.gravity);
      const weighted = dist.map(([k, p]) => {
        const rootPc = mod12(root + Number(k[0]));
        const factor = isDiatonicMemo(rootPc, String(k[1]), ctx.tonic, ctx.mode, ctx.home) ? bonus : 1.0;
        return [k, p * factor];
      });
      let total = 0;
      for (const [, p] of weighted) total += p;
      if (total > 0) {
        dist = weighted.map(([k, p]) => [k, p / total]);
      }
    }
    const chosen = pick(rngFn, dist);
    if (!chosen) return [5, "7"];
    return [Number(chosen[0]), String(chosen[1])];
  }

  /**
   * drawDuration(quality, mpos, rng, maxD) -> int beats.
   *
   * Learned harmonic rhythm D(d | quality, metric position). Backoff chain:
   * duration["q|m"] -> duration_backoff[m] -> duration_all. `maxD` truncates
   * the support so a run lands exactly on target; 1 beat is always present so a
   * truncated draw always exists.
   */
  function drawDuration(quality, mpos, rngFn, maxD) {
    const chain = [dur[quality + "|" + mpos], durBackoff.get(mpos), durAll];
    for (let level = 0; level < chain.length; level++) {
      let dist = chain[level];
      if (!dist || dist.length === 0) continue;
      if (maxD != null) {
        dist = dist.filter(([d]) => d <= maxD);
        if (dist.length === 0) continue;
        let total = 0;
        for (const [, p] of dist) total += p;
        dist = dist.map(([d, p]) => [d, p / total]);
      }
      const d = pick(rngFn, dist);
      if (d) return Math.trunc(d);
    }
    return maxD == null ? 1 : Math.min(1, maxD);
  }

  /**
   * run(root, quality, target, rng, truncate, ctx) -> segs | null.
   *
   * Walk the chain from (root, quality) for exactly `target` beats. Returns
   * [[rootPc, quality, dur], ...], or null if it overshot (only possible when
   * truncate=false). Metric position is tracked so durations stay idiomatic.
   */
  function run(root, quality, target, rngFn, truncate, ctx) {
    const segs = [];
    let onset = 0;
    let q = quality;
    let r = root;
    while (onset < target) {
      const remaining = target - onset;
      const d = drawDuration(q, onset % 4, rngFn, truncate ? remaining : null);
      if (d > remaining) return null; // overshot: reject this run
      segs.push([r, q, d]);
      onset += d;
      if (onset === target) return segs;
      const [dpc, nextQ] = drawMotion(q, rngFn, r, ctx);
      q = nextQ;
      r = mod12(r + dpc);
    }
    return segs;
  }

  /**
   * body(root, quality, target, rng, ctx) -> segs.
   * Exact-length body: rejection-sample for distributional fidelity, then fall
   * back to a truncated draw that cannot fail (never returns silence).
   */
  function body(root, quality, target, rngFn, ctx) {
    if (target <= 0) return [];
    for (let i = 0; i < MAX_TRIES; i++) {
      const r = run(root, quality, target, rngFn, false, ctx);
      if (r) return r;
    }
    return run(root, quality, target, rngFn, true, ctx) || [[root, quality, target]];
  }

  /**
   * dominantDuration(avail, rng) -> int beats.
   *
   * How long the cadential V is held. A V of length d starts at metric position
   * m = (avail - d) % 4, so each candidate length implies a DIFFERENT metric
   * position and candidates must be scored by the JOINT P(m) * P(d | q='7', m),
   * NOT P(d | q, m) alone: an off-beat onset is almost always 1 beat, so the
   * conditional would make a weak-beat 1-beat dominant look near-certain and the
   * cadence would land off the beat.
   */
  function dominantDuration(avail, rngFn) {
    const maxD = avail > 1 ? Math.max(1, Math.min(avail - 1, 4)) : 1;
    const weights = [];
    for (const d of support) {
      if (d > maxD) continue;
      const mpos = mod12Metric(avail - d);
      const pM = onsetPos.get(mpos) || 0;
      if (pM <= 0) continue;
      const chain = [dur["7|" + mpos], durBackoff.get(mpos), durAll];
      for (let level = 0; level < chain.length; level++) {
        const dist = chain[level];
        if (dist) {
          let pD = 0;
          for (const [dd, prob] of dist) {
            if (dd === d) { pD = prob; break; }
          }
          if (pD > 0) weights.push([d, pM * pD]);
          break; // consult only the first available backoff level
        }
      }
    }
    if (weights.length === 0) return maxD;
    let total = 0;
    for (const [, p] of weights) total += p;
    const normalized = weights.map(([d, p]) => [d, p / total]);
    const chosen = pick(rngFn, normalized);
    return chosen != null ? Math.trunc(chosen) : maxD;
  }

  // Metric position is onset % 4; avail-d is always >= 0 here, but route it
  // through a mod-4 that is safe for the general case for parity with Python.
  function mod12Metric(x) {
    return ((x % 4) + 4) % 4;
  }

  /** parseSymbol('D:min7') -> [pc, quality] | null (quality must be in model). */
  function parseSymbol(sym) {
    if (!sym || sym.indexOf(":") === -1) return null;
    const i = sym.indexOf(":");
    const root = sym.slice(0, i);
    const quality = sym.slice(i + 1);
    const pc = PITCH_CLASSES[root];
    if (pc == null || qualities.indexOf(quality) === -1) return null;
    return [pc, quality];
  }

  /**
   * render(segs) -> [[symbol, dur], ...].
   * Pitch classes -> canonical runtime spelling ('Bb:maj7', never 'B-').
   * COALESCE adjacent identical symbols by summing durations: splicing the
   * cadential V onto a body that already ended on the V would otherwise
   * re-trigger the chord instead of sustaining it.
   */
  function render(segs) {
    const out = [];
    for (const [r, q, d] of segs) {
      const symbol = CANON_ROOT[mod12(r)] + ":" + q;
      if (out.length && out[out.length - 1][0] === symbol) {
        out[out.length - 1][1] += Math.trunc(d);
      } else {
        out.push([symbol, Math.trunc(d)]);
      }
    }
    return out;
  }

  /**
   * generate(bars, key, opts) -> [[chordSymbol, durationBeats], ...].
   *
   * Durations sum to exactly bars*4. `cadence` (0..1) is BOTH the probability
   * the final bar resolves via an authentic V(7)->I and the tonal-gravity
   * strength keeping the body's roots in key. `seedChord` optionally starts the
   * walk off the tonic; a numeric `seed` makes it deterministic.
   */
  function generate(bars, key, opts) {
    const o = opts || {};
    const seedChord = o.seedChord != null ? o.seedChord : null;
    const cadence = o.cadence != null ? o.cadence : 1.0;
    const seed = o.seed != null ? o.seed : null;

    let barsInt = Math.trunc(Number(bars));
    if (!Number.isFinite(barsInt)) barsInt = 1;
    barsInt = Math.max(1, barsInt);
    const total = barsInt * 4;
    const rngFn = seed != null ? makeRng(seed) : rng;

    let { tonicPc, mode } = parseKey(key);
    if (tonicPc == null) tonicPc = 0;
    const cad = cadence_for(mode);
    const finalQ = String((cad && cad.final) || (mode === "min" ? "min7" : "maj7"));
    const preQ = String((cad && cad.pre) || "7");
    const gravity = clamp01(cadence);
    const ctx = {
      tonic: tonicPc,
      mode: mode,
      gravity: gravity,
      home: DIATONIC[mode] || DIATONIC.maj,
    };

    let startRoot = tonicPc;
    let startQ = finalQ;
    if (seedChord) {
      const parsed = parseSymbol(seedChord);
      if (parsed) {
        startRoot = parsed[0];
        startQ = parsed[1];
      }
    }

    const resolve = rngFn() < gravity;
    if (!resolve || total < 4 + 2) {
      // Free ending: one unconstrained run of the whole phrase.
      return render(body(startRoot, startQ, total, rngFn, ctx));
    }

    // Reserve the final bar for the tonic, reached by a V(7). The V's duration
    // is drawn at the metric position it actually lands on, so a mid-bar
    // dominant fills to the downbeat like the corpus does.
    const tail = CADENCE_BARS * 4;
    const domRoot = mod12(tonicPc + 7); // V: a descending fifth from the tonic
    const domDur = dominantDuration(total - tail, rngFn);
    const bodyTarget = total - tail - domDur;

    const segs = bodyTarget > 0 ? body(startRoot, startQ, bodyTarget, rngFn, ctx) : [];
    segs.push([domRoot, preQ, domDur]);
    segs.push([tonicPc, finalQ, tail]);
    return render(segs);
  }

  function cadence_for(mode) {
    const key = mode === "min" ? "minor" : "major";
    return cadence[key] || {};
  }

  return { generate };
}

module.exports = {
  createPhraseEngine,
  DIATONIC,
  MAX_TONAL_PULL,
  isDiatonic,
};

File: python/src/engines/phrase_engine.py
"""Phrase generation: the SEQUENCE is the unit of analysis.

Answers "give me an N-bar phrase in key K" with a list of ``(chord, duration)``
pairs whose harmonic rhythm is **learned** and whose ending is a cadence —
rather than the chord-at-a-time walk (``/chord/input``) driven by hand-authored
rhythm templates.

Model: an observed-state, explicit-duration (semi-Markov) chain over ROOT
MOTION, trained by ``scripts/build_phrase_model.py``. Three decisions define it:

  1. **Explicit durations.** An ordinary Markov chain gives each chord a
     geometric dwell time; it cannot express "chords change on the downbeat or
     at mid-bar". A per-state duration histogram conditioned on metric position
     — ``D(d | quality, onset mod 4)`` — can, and that is the harmonic rhythm.
     (In the corpus, a chord starting on beat 2 lasts exactly 2 beats 95% of
     the time; one starting on a downbeat usually lasts 4.)

  2. **No key is estimated.** The training corpus modulates constantly, so any
     key-relative "degree" alphabet inherits the key-finder's errors into every
     label. The state is instead the interval from the previous root plus the
     new quality, which needs no key and covers the data completely. The
     caller's key is the single absolute anchor, applied only at generation.

  3. **No hidden state, therefore no EM.** Once the representation is chosen,
     every state is observed; training is counting with smoothing. There is no
     Baum-Welch here, and none is needed.

The cadence is *imposed*, not hoped for: the last bar is reserved for the tonic
and reached by an authentic V(7) -> I motion. (The corpus itself resolves only
~40% of the time, so waiting for a cadence to emerge would not work.)

Imports only ``random`` + ``chord_vocab`` — no torch, no music21, no numpy — so
the service keeps its ~0.12 s cold start.
"""

from __future__ import annotations

import json
import random
from pathlib import Path

from ..chord_vocab import CANON_ROOT, PITCH_CLASSES, QUALITY_INTERVALS, parse_key

# Duration draws are rejection-sampled for exact phrase length; after this many
# unlucky runs we switch to a truncated draw that always fits (see _run).
MAX_TRIES = 200
# Beats reserved at the end for the tonic resolution.
CADENCE_BARS = 1
# Scale degrees (semitones from the tonic) treated as "at home" in each mode.
# Minor includes the raised 7th so the dominant belongs.
DIATONIC = {
    "maj": frozenset({0, 2, 4, 5, 7, 9, 11}),
    "min": frozenset({0, 2, 3, 5, 7, 8, 10, 11}),
}
# How much a fully-open Cadence dial favors in-key roots. Multiplicative on the
# learned transition weight, so at gravity=0 the corpus is reproduced exactly.
MAX_TONAL_PULL = 4.0


class PhraseModelError(RuntimeError):
    """The trained model file is missing or malformed."""


def _pick(rng: random.Random, dist: list) -> object:
    """Sample a key from a [[key, prob], ...] table (probabilities sum to ~1)."""
    r = rng.random()
    acc = 0.0
    for key, p in dist:
        acc += p
        if r <= acc:
            return key
    return dist[-1][0] if dist else None


class PhraseEngine:
    """Generate whole phrases: [(chord_symbol, duration_in_beats), ...]."""

    def __init__(self, model_path: str | Path, seed: int | None = None) -> None:
        path = Path(model_path)
        try:
            m = json.loads(path.read_text())
        except (OSError, ValueError) as exc:
            raise PhraseModelError(f"cannot load phrase model {path}: {exc}") from exc
        for required in ("transition", "duration", "cadence", "duration_support"):
            if required not in m:
                raise PhraseModelError(f"phrase model {path} missing '{required}'")
        self._m = m
        self._rng = random.Random(seed)
        # transition tables come back from JSON as [[dpc, quality], p]
        self._trans: dict[str, list] = m["transition"]
        self._trans_marginal: list = m["transition_marginal"]
        self._dur: dict[str, list] = m["duration"]
        self._dur_backoff: dict[str, list] = m["duration_backoff"]
        self._dur_all: list = m["duration_all"]
        self._onset_pos: dict[int, float] = {int(k): p for k, p in m.get("onset_position", [])}
        self._cadence: dict = m["cadence"]
        self._diatonic_memo: dict[tuple[int, str, str], bool] = {}

    def _is_diatonic(self, root: int, quality: str, tonic: int, mode: str,
                     home: frozenset[int]) -> bool:
        """Are ALL of this chord's notes in the key's scale?

        Testing the root alone is not enough: in A minor, A:maj7 has a diatonic
        ROOT but a raised third and seventh. Testing the chord's pitch classes
        yields exactly the diatonic sevenths of the key (in C: Cmaj7 Dm7 Em7
        Fmaj7 G7 Am7 Bhdim7) and correctly rejects A:maj7 in A minor.
        """
        degree = (root - tonic) % 12
        memo_key = (degree, quality, mode)
        hit = self._diatonic_memo.get(memo_key)
        if hit is None:
            intervals = QUALITY_INTERVALS.get(quality)
            hit = bool(intervals) and all((degree + iv) % 12 in home for iv in intervals)
            self._diatonic_memo[memo_key] = hit
        return hit

    # --- distribution access with backoff ---------------------------------

    def _draw_motion(self, quality: str, rng: random.Random, root: int = 0,
                     ctx: dict | None = None) -> tuple[int, str]:
        """(root motion in semitones, next quality) given the current quality.

        TONAL GRAVITY: the model is trained key-free, so on its own an order-1
        motion chain reproduces the corpus's constant modulation and a long
        phrase drifts far from home. Since the key IS known at generation time,
        `ctx['gravity']` (>0) reweights candidates toward chords that are wholly
        diatonic in it. The bonus is multiplicative, so out-of-key colour (a
        secondary dominant, a tritone sub) stays reachable — it just gets rarer.
        At gravity=0 the corpus distribution is reproduced untouched.
        """
        dist = self._trans.get(quality) or self._trans_marginal
        if not dist:
            return 5, "7"  # descending fifth into a dominant: the corpus's backbone
        if ctx and ctx["gravity"] > 0:
            bonus = 1.0 + MAX_TONAL_PULL * min(1.0, ctx["gravity"])
            tonic, mode, home = ctx["tonic"], ctx["mode"], ctx["home"]
            weighted = [
                [k, p * (bonus if self._is_diatonic((root + int(k[0])) % 12, str(k[1]),
                                                    tonic, mode, home) else 1.0)]
                for k, p in dist
            ]
            total = sum(p for _, p in weighted)
            if total > 0:
                dist = [[k, p / total] for k, p in weighted]
        pick = _pick(rng, dist)
        if not pick:
            return 5, "7"
        return int(pick[0]), str(pick[1])

    def _draw_duration(self, quality: str, m: int, rng: random.Random,
                       max_d: int | None = None) -> int:
        """Learned harmonic rhythm: duration given quality + metric position.

        `max_d` truncates the support so a run can land exactly on its target;
        1 beat is always in the support, so a truncated draw always exists.
        """
        for dist in (self._dur.get(f"{quality}|{m}"), self._dur_backoff.get(str(m)), self._dur_all):
            if not dist:
                continue
            if max_d is not None:
                dist = [[d, p] for d, p in dist if d <= max_d]
                if not dist:
                    continue
                total = sum(p for _, p in dist)
                dist = [[d, p / total] for d, p in dist]
            d = _pick(rng, dist)
            if d:
                return int(d)
        return 1 if max_d is None else min(1, max_d)

    # --- run construction ---------------------------------------------------

    def _run(self, root: int, quality: str, target: int, rng: random.Random,
             truncate: bool, ctx: dict) -> list[tuple[int, str, int]] | None:
        """Walk the chain from (root, quality) for exactly `target` beats.

        Returns [(root_pc, quality, dur), ...] or None if it overshot (only
        possible when truncate=False). Metric position is tracked so durations
        stay idiomatic (mid-bar chords fill to the downbeat).
        """
        segs: list[tuple[int, str, int]] = []
        onset = 0
        while onset < target:
            remaining = target - onset
            d = self._draw_duration(quality, onset % 4, rng, remaining if truncate else None)
            if d > remaining:
                return None  # overshot: reject this run
            segs.append((root, quality, d))
            onset += d
            if onset == target:
                return segs
            dpc, quality = self._draw_motion(quality, rng, root, ctx)
            root = (root + dpc) % 12
        return segs

    def _body(self, root: int, quality: str, target: int,
              rng: random.Random, ctx: dict) -> list[tuple[int, str, int]]:
        """Exact-length body. Rejection-sample for distributional fidelity, then
        fall back to a truncated draw that cannot fail (never returns silence)."""
        if target <= 0:
            return []
        for _ in range(MAX_TRIES):
            run = self._run(root, quality, target, rng, False, ctx)
            if run:
                return run
        return self._run(root, quality, target, rng, True, ctx) or [(root, quality, target)]

    # --- public API ---------------------------------------------------------

    def generate(self, bars: int, key: str, seed_chord: str | None = None,
                 cadence: float = 1.0, seed: int | None = None) -> list[tuple[str, int]]:
        """An `bars`-bar phrase in `key` as [(chord, duration_beats), ...].

        Durations sum to exactly ``bars * 4``, and their pattern is the learned
        harmonic rhythm.

        `cadence` (0..1) is the phrase's pull toward home, and does two things:
        it is the probability that the final bar is a tonic reached by an
        authentic V(7) -> I, and it is the strength of the tonal gravity that
        keeps the body's roots in key. At 0 the phrase wanders and ends
        unresolved (as ~60% of the corpus does); at 1 it stays home and
        cadences. `seed_chord` optionally starts the walk somewhere other than
        the tonic. `seed` makes the result deterministic.
        """
        bars = max(1, int(bars))
        total = bars * 4
        rng = random.Random(seed) if seed is not None else self._rng

        tonic_pc, mode = parse_key(key)
        if tonic_pc is None:
            tonic_pc = 0
        cad = self._cadence.get("minor" if mode == "min" else "major", {})
        final_q = str(cad.get("final") or ("min7" if mode == "min" else "maj7"))
        pre_q = str(cad.get("pre") or "7")
        gravity = max(0.0, min(1.0, float(cadence)))
        ctx = {"tonic": tonic_pc, "mode": mode, "gravity": gravity,
               "home": DIATONIC.get(mode, DIATONIC["maj"])}

        start_root, start_q = tonic_pc, final_q
        if seed_chord:
            parsed = self._parse_symbol(seed_chord)
            if parsed:
                start_root, start_q = parsed

        resolve = rng.random() < gravity
        if not resolve or total < 4 + 2:
            # Free ending: one unconstrained run of the whole phrase.
            return self._render(self._body(start_root, start_q, total, rng, ctx))

        # Reserve the final bar for the tonic, and the V(7) that leads to it.
        # The V's duration is drawn at the metric position it actually lands on,
        # so a mid-bar dominant fills to the downbeat like the corpus does.
        tail = CADENCE_BARS * 4
        dom_root = (tonic_pc + 7) % 12  # V: a descending fifth from the tonic
        dom_dur = self._dominant_duration(total - tail, rng)
        body_target = total - tail - dom_dur

        segs = self._body(start_root, start_q, body_target, rng, ctx) if body_target > 0 else []
        segs.append((dom_root, pre_q, dom_dur))
        segs.append((tonic_pc, final_q, tail))
        return self._render(segs)

    def _dominant_duration(self, avail: int, rng: random.Random) -> int:
        """How long the cadential V is held. Must leave >=1 beat of body when
        there is room for one, so short phrases still start somewhere.

        A V of length d starts at beat (avail - d), i.e. at metric position
        m = (avail - d) % 4. Each candidate length therefore implies a DIFFERENT
        metric position, so candidates must be scored by the joint

            P(m) * P(d | quality='7', m)

        and NOT by P(d | q, m) alone: an off-beat onset is almost always 1 beat
        long, so the conditional would make a weak-beat, 1-beat dominant look
        near-certain and the cadence would land off the beat.
        """
        max_d = max(1, min(avail - 1, 4)) if avail > 1 else 1
        weights: list[list] = []
        for d in self._m["duration_support"]:
            if d > max_d:
                continue
            m = (avail - d) % 4
            p_m = self._onset_pos.get(m, 0.0)
            if p_m <= 0:
                continue
            for dist in (self._dur.get(f"7|{m}"), self._dur_backoff.get(str(m)), self._dur_all):
                if dist:
                    p_d = next((prob for dd, prob in dist if dd == d), 0.0)
                    if p_d > 0:
                        weights.append([d, p_m * p_d])
                    break
        if not weights:
            return max_d
        total = sum(p for _, p in weights)
        return int(_pick(rng, [[d, p / total] for d, p in weights]) or max_d)

    def _parse_symbol(self, sym: str) -> tuple[int, str] | None:
        if not sym or ":" not in sym:
            return None
        root, _, quality = sym.partition(":")
        pc = PITCH_CLASSES.get(root)
        if pc is None or quality not in self._m.get("qualities", []):
            return None
        return pc, quality

    @staticmethod
    def _render(segs: list[tuple[int, str, int]]) -> list[tuple[str, int]]:
        """Pitch classes -> canonical runtime spelling ('Bb:maj7', never 'B-').

        Coalesces adjacent identical chords into one held chord. Training data
        is run-length encoded, so the chain itself never repeats a chord — but
        splicing the cadential V onto a body that already ended on the V would,
        and the device would re-trigger the chord instead of sustaining it.
        """
        out: list[tuple[str, int]] = []
        for r, q, d in segs:
            symbol = f"{CANON_ROOT[r % 12]}:{q}"
            if out and out[-1][0] == symbol:
                out[-1] = (symbol, out[-1][1] + int(d))
            else:
                out.append((symbol, int(d)))
        return out
